### PR TITLE
Autoscaling & Tagging features

### PR DESCRIPTION
**Pull request by gwdp**
This PR allows auto-scaling per service, resources tagging, output service cluster endpoints and fixes deletion policy.  Changes made on #5 are used as the foundation. 

Have fun!

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
-  "name": "serverless-fargate-plugin",
-  "version": "1.0.11",
+  "name": "@ikonintegration/serverless-fargate-plugin",
```
Comment (oscarwest) [inline package.json]:
```suggestion
  "name": "serverless-fargate-plugin",
```

File: src/index.ts [hunk 2]
```diff
@@ -17,19 +17,35 @@ class ServerlessFargatePlugin {
...
-
-        // merge all our stuff into resources
-        Object.assign(
-            this.serverless.service.provider.compiledCloudFormationTemplate.Resources,
-            vpc.generate(),
-            cluster.generate()
-        );
+        
+        //No cluster section specified, don't process
+        if (!options || !options.length) return;
```
Comment (oscarwest) [inline src/index.ts]: Maybe add some console.log's here for clarity

**Review by oscarwest — COMMENTED** [review]
Nice work, @honerlaw how can we get changes deployed faster in the future?

File: package.json [hunk 3]
```diff
@@ -1,6 +1,6 @@
...
+  "version": "1.0.63",
```
Comment (honerlaw) [inline package.json:3]: This shouldn't be changed.

File: src/index.ts [hunk 4]
```diff
@@ -17,19 +17,35 @@ class ServerlessFargatePlugin {
...
+
+        //For each cluster
+        for (let clusterOption of options) {
+            if (clusterOption && clusterOption.vpc) { //sanity check for empty objects
+                const index = options.indexOf(clusterOption);
```
Comment (honerlaw) [inline src/index.ts]: if we need the index, we should just use basic for loop instead of for/of

File: src/options.ts [hunk 5]
```diff
@@ -1,18 +1,50 @@
...
+    cooldownOut?: number; //defaults to cooldown but has priority over it
+    targetValue: number;
+}
+
 export interface IServiceOptions {
     name: string;
     cpu: number;
     memory: number;
-    port: number;
+    port?: number;
```
Comment (honerlaw) [inline src/options.ts]: Probably should add a comment saying it defaults to a randomly generated port.

File: src/resource.ts [hunk 6]
```diff
@@ -54,4 +70,8 @@ export abstract class Resource<T> {
         return this.namePrefix;
     }
 
-}
\ No newline at end of file
+    public getOutputs(): any {
```
Comment (honerlaw) [inline src/resource.ts]: could this just be `public abstract getOutputs(): any`? Let the implementing class denote if its an empty object or not.

File: src/resources/cluster.ts [hunk 7]
```diff
@@ -1,107 +1,172 @@
...
+                            "Ref": this.getName(NamePostFix.CONTAINER_SECURITY_GROUP)
+                        },
+                        "IpProtocol": -1,
+                        "SourceSecurityGroupId": {
+                            "Ref": ELBServiceSecGroup
+                        }
+                    }
+                }
+            } : {
+                /*TODO: if not public AND also not specifiying a VPC, different secgroup must be created*/
```
Comment (honerlaw) [inline src/resources/cluster.ts:168]: We should console.err this or something, so its very obvious that this is not being created.

**Review by honerlaw — CHANGES_REQUESTED** [review]
couple of small things

File: package.json (hunk 3 above)
Comment (gwdp) [reply package.json:3]: My bad, changing it now.

File: src/index.ts (hunk 4 above)
Comment (gwdp) [reply src/index.ts]: Unused index, removing it.

File: src/index.ts (hunk 2 above)
Comment (gwdp) [reply src/index.ts]: Done

File: src/options.ts (hunk 5 above)
Comment (gwdp) [reply src/options.ts]: Done :)

File: src/resource.ts (hunk 6 above)
Comment (gwdp) [reply src/resource.ts]: Awesome, done as well!

**Comment by gwdp** [comment]
Agreed with all the suggestions. Changed and pushed :)  

Let me know if you guys have any other suggestions or comments on it.

**Review by honerlaw — APPROVED** [review]

**Pull request merged**